typing/typecore.ml

```diff
@@ -4437,14 +4492,14 @@ and type_application env (mode : Alloc_mode.t) funct sargs =
   match sargs with
   | (* Special case for ignore: avoid discarding warning *)
     [Nolabel, sarg] when is_ignore funct ->
-      let ty_arg, ty_res = filter_arrow env (instance funct.exp_type) Nolabel in
-      (* FIXME arg mode *)
-      let exp = type_expect env mode sarg (mk_expected ty_arg) in
+      let marg, ty_arg, _mres, ty_res =
+        filter_arrow env (instance funct.exp_type) Nolabel in
+      let exp = type_expect env marg sarg (mk_expected ty_arg) in
       check_partial_application false exp;
       ([Nolabel, Some exp], ty_res)
   | _ ->
     let ty = funct.exp_type in
-    type_args [] ty (instance ty) sargs
+    type_args [] ty (instance ty) sargs Alloc_heap
```

Comment:
This should be `funct.exp_mode`, otherwise omitting the first parameter allows you to leak a local function:
```ocaml
# let foo () =
    let[@stack] s = "hello" in
    let[@stack] bar =
      fun ~a ~b -> print s
    in
    bar ~b:();;
val foo : unit -> a:'a -> unit = <fun>
```

Reply:
Ostensibly yes! The issue is that functions are typechecked in mode local, and using exp_mode here means that partial applications that do not apply the first argument count as local, even if everything involved is heap.

The fix is to change the typing rule for applications to typecheck the function with a fresh mode variable, rather than at `Alloc_local`. (This fresh mode variable is usually useless but can leak out in some partial application cases). Of course, that fix is waiting for mode variables.

The hack here was to type the functions at mode local but use Alloc_heap in type_application, which as you point out is unsound. I'll change it to do the right thing here but type functions at Alloc_heap, which will be sound but breaks local functions (until mode variables appear).